File: go/tests/social_board_test.go
// social_board_test.go
//
// Verifies the CircleAI.Social port (social_board.go): post/get, reaction counts
// (case-insensitive), follow/unfollow with self-follow rejection, feed
// newest-first, and followers listing.

package circleai_test

import (
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestSocial_ReactionsAndFollow(t *testing.T) {
	b := circleai.NewInMemorySocialBoard()
	now := time.Now().UTC()
	b.Post(circleai.SocialPost{PostId: "p1", AuthorId: "u2", Body: "hi", AtUtc: now, Tags: []string{"greeting"}})
	if got, ok := b.GetPost("p1"); !ok || got.Body != "hi" {
		t.Fatalf("get post = %+v ok=%v", got, ok)
	}
	b.React(circleai.Reaction{PostId: "p1", UserId: "u1", Kind: "like", AtUtc: now})
	b.React(circleai.Reaction{PostId: "p1", UserId: "u3", Kind: "LIKE", AtUtc: now})
	b.React(circleai.Reaction{PostId: "p1", UserId: "u4", Kind: "love", AtUtc: now})
	if n := b.ReactionCount("p1", "like"); n != 2 {
		t.Fatalf("reaction count = %d, want 2", n)
	}

	if err := b.Follow(circleai.Follow{FollowerId: "u1", FolloweeId: "u1", AtUtc: now}); err == nil {
		t.Fatalf("self-follow must error")
	}
	if err := b.Follow(circleai.Follow{FollowerId: "u1", FolloweeId: "u2", AtUtc: now}); err != nil {
		t.Fatalf("follow: %v", err)
	}
	if f := b.Followers("u2"); len(f) != 1 || f[0] != "u1" {
		t.Fatalf("followers failed: %+v", f)
	}
	b.Unfollow("u1", "u2")
	if f := b.Followers("u2"); len(f) != 0 {
		t.Fatalf("followers after unfollow failed: %+v", f)
	}
}

func TestSocial_Feed(t *testing.T) {
	b := circleai.NewInMemorySocialBoard()
	now := time.Now().UTC()
	b.Follow(circleai.Follow{FollowerId: "u1", FolloweeId: "u2", AtUtc: now})
	b.Post(circleai.SocialPost{PostId: "p1", AuthorId: "u2", Body: "old", AtUtc: now.Add(-time.Hour)})
	b.Post(circleai.SocialPost{PostId: "p2", AuthorId: "u2", Body: "new", AtUtc: now})
	b.Post(circleai.SocialPost{PostId: "p3", AuthorId: "u9", Body: "unfollowed", AtUtc: now})

	feed := b.FeedFor("u1", 20)
	if len(feed) != 2 || feed[0].PostId != "p2" || feed[1].PostId != "p1" {
		t.Fatalf("feed newest-first (followed only) failed: %+v", feed)
	}
	if lim := b.FeedFor("u1", 1); len(lim) != 1 || lim[0].PostId != "p2" {
		t.Fatalf("feed limit failed: %+v", lim)
	}
}

File: typescript/tests/social_board.test.ts
// social_board.test.ts
// Verifies the CircleAI.Social port: reaction counts, follow/unfollow guards,
// follow-graph feed newest-first + limit, followers.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  InMemorySocialBoard,
  SocialDomainContext,
  socialPost,
  reaction,
  follow,
} from "../src/social/index";

describe("InMemorySocialBoard", () => {
  it("counts reactions by kind case-insensitively", () => {
    const b = new InMemorySocialBoard();
    b.post(socialPost("p1", "u1", "hi", new Date("2026-01-01T00:00:00Z"), []));
    b.react(reaction("p1", "u2", "Like", new Date("2026-01-01T00:00:00Z")));
    b.react(reaction("p1", "u3", "like", new Date("2026-01-01T00:00:00Z")));
    b.react(reaction("p1", "u4", "love", new Date("2026-01-01T00:00:00Z")));
    assert.equal(b.reactionCount("p1", "LIKE"), 2);
    assert.equal(b.reactionCount("p1", "love"), 1);
    assert.equal(b.getPost("p1")?.body, "hi");
  });

  it("prevents following yourself and supports unfollow", () => {
    const b = new InMemorySocialBoard();
    assert.throws(() => b.follow(follow("u1", "u1", new Date())), /Cannot follow yourself/);
    b.follow(follow("u1", "u2", new Date("2026-01-01T00:00:00Z")));
    b.follow(follow("u1", "u3", new Date("2026-01-01T00:00:00Z")));
    b.unfollow("u1", "u2");
    assert.deepEqual(b.followers("u3"), ["u1"]);
    assert.deepEqual(b.followers("u2"), []);
  });

  it("builds a feed of followed authors' posts, newest-first, capped by limit", () => {
    const b = new InMemorySocialBoard();
    b.follow(follow("u1", "a", new Date("2026-01-01T00:00:00Z")));
    b.follow(follow("u1", "b", new Date("2026-01-01T00:00:00Z")));
    b.post(socialPost("p1", "a", "1", new Date("2026-01-01T00:00:00Z"), []));
    b.post(socialPost("p2", "b", "2", new Date("2026-01-03T00:00:00Z"), []));
    b.post(socialPost("p3", "c", "3", new Date("2026-01-04T00:00:00Z"), [])); // not followed
    assert.deepEqual(
      b.feedFor("u1").map((p) => p.postId),
      ["p2", "p1"],
    );
    assert.deepEqual(
      b.feedFor("u1", 1).map((p) => p.postId),
      ["p2"],
    );
  });

  it("feedFor throws on non-positive limit", () => {
    const b = new InMemorySocialBoard();
    assert.throws(() => b.feedFor("u1", 0), /limit/);
  });

  it("domain context exposes prompt + compliance + tools", () => {
    assert.ok(SocialDomainContext.systemPromptSnippet.includes("[DOMAIN: Social]"));
    assert.deepEqual(SocialDomainContext.complianceFlags, [
      "POPIA",
      "ASA_Advertising_Code",
      "Platform_Community_Standards",
    ]);
    assert.deepEqual(SocialDomainContext.suggestedTools, ["social_media_api", "analytics", "content_planner", "image_tools"]);
  });
});
